---
title: Home 
---

<h1>SeqWare</h1>

<p>The open source SeqWare project is a portable software infrastructure designed to analyze massive genomics datasets produced by contemporary and emerging technologies, in particular Next Generation Sequencing (NGS) platforms.  It consists of a comprehensive suite of tools focused on enabling the end-to-end analysis of sequence data -- from from raw base calling to analyzed variants ready for interpretation by users.</p>


<table>
<tr>
<th>Users</th>
<th>Administrators</th>
<th>Developers</th>
</tr>
<tr>
<td><img src="pictures/Use.png"/></td>
<td><img src="pictures/Deploy.png"/></td>
<td><img src="pictures/Develop.png"/></td>
</tr>
<tr>
<td width="200">Create and run workflows either using a web application or command line interface</td>
<td width="200">Deploy SeqWare on a virtual machine, in the Cloud, or on your own cluster</td>
<td width="200">Contribute to SeqWare by getting involved with core development</td>
</tr>
</table>


SeqWare is currently used by a variety of NGS users including the <a href="http://unclineberger.org/">Lineberger Comprehensive Cancer Center at UNC</a>, at the <a href="http://oicr.on.ca/">Ontario Institute for Cancer Research in Toronto</a> and by our commercial partner <a href="http://nimbusinformatics.com/">Nimbus Informatics.</a></p>

