<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_34) on Mon Oct 22 09:23:56 EDT 2012 -->
<TITLE>
Uses of Class net.sourceforge.seqware.pipeline.module.Module
</TITLE>

<META NAME="date" CONTENT="2012-10-22">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Uses of Class net.sourceforge.seqware.pipeline.module.Module";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?net/sourceforge/seqware/pipeline/module//class-useModule.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Module.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>net.sourceforge.seqware.pipeline.module.Module</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Packages that use <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.module"><B>net.sourceforge.seqware.pipeline.module</B></A></TD>
<TD>This package provides the SeqWare module interface and API, including
 common code and classes.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules"><B>net.sourceforge.seqware.pipeline.modules</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.alignment"><B>net.sourceforge.seqware.pipeline.modules.alignment</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.alignment.bwa"><B>net.sourceforge.seqware.pipeline.modules.alignment.bwa</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.BamFilters"><B>net.sourceforge.seqware.pipeline.modules.BamFilters</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.examples"><B>net.sourceforge.seqware.pipeline.modules.examples</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.io"><B>net.sourceforge.seqware.pipeline.modules.io</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.qc"><B>net.sourceforge.seqware.pipeline.modules.qc</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.quantification"><B>net.sourceforge.seqware.pipeline.modules.quantification</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.queryengine"><B>net.sourceforge.seqware.pipeline.modules.queryengine</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.modules.utilities"><B>net.sourceforge.seqware.pipeline.modules.utilities</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#net.sourceforge.seqware.pipeline.runner"><B>net.sourceforge.seqware.pipeline.runner</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.module"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/package-summary.html">net.sourceforge.seqware.pipeline.module</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/package-summary.html">net.sourceforge.seqware.pipeline.module</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/TestModule.html" title="class in net.sourceforge.seqware.pipeline.module">TestModule</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is an empty module mainly for testing purpose.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/package-summary.html">net.sourceforge.seqware.pipeline.modules</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/package-summary.html">net.sourceforge.seqware.pipeline.modules</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/bmfsplit.html" title="class in net.sourceforge.seqware.pipeline.modules">bmfsplit</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/GenericCommandRunner.html" title="class in net.sourceforge.seqware.pipeline.modules">GenericCommandRunner</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is a very simple module that takes a few options on the command line and
 then passes the rest of the arguments to whatever command is specified.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/GenericMetadataSaver.html" title="class in net.sourceforge.seqware.pipeline.modules">GenericMetadataSaver</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is a very simple module is used to save both a processing event and 0 or
 more files to the metadb.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/Picard.html" title="class in net.sourceforge.seqware.pipeline.modules">Picard</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/SamTools.html" title="class in net.sourceforge.seqware.pipeline.modules">SamTools</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;FIXME: if creating pileup requres a .fai file!</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/SNVMix.html" title="class in net.sourceforge.seqware.pipeline.modules">SNVMix</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>
 Syntax:
        ./SNVMix2       -m <modelfile> [-i <infile>] [-o <outfile>] [-T | -C | -F] [-p < q | m | s >] [-t < mb | m | b | M | Mb | MB | SNVMix1>] [-q <#>] [-Q <#>] [-a/-b/-d <#,#,#>] [-M <trainP file>] [-h]
        Required:
        -m <modelfile>  model file, a line for mu and a line for pi, three
                        space-separated values each, like:
                        mu 0.xxxxxxxx 0.xxxxxxxx 0.xxxxxxxx
                        pi 0.xxxxxxxx 0.xxxxxxxx 0.xxxxxxxx
        Optional:
        -i <infile>     quality pileup, from pileup2matlab.pl script (def: STDIN)
        -o <outfile>    where to put the results (def: STDOUT)
        -T | -C | -F    Train, Classify or Filter (def: Classify)
        -p q|m|s        Input pileup format (def: s)
                        q = quality
                        m = MAQ
                        s = SAMtools
        -t mb|m|b|M|Mb|MB|SNVMix1
                        Filter type (def: mb)
                        mb      Lowest between map and base quality
                        m       Filter on map, and use as surrogate for base quality
                        b       Filter on base quality, take map quality as 1
                        M       Filter on map quality but use only base quality
                        Mb      Filter on map quality and use both map and base qualities
                        MB      Filter on map quality AND base quality
                        SNVMix1 Filter on base quality and map quality, afterwards consider them perfect
        -q #            Cutoff Phred value for Base Quality, anything <= this value is ignored (def: Q19)
        -Q #            Cutoff Phred value for Map Quality, anything <= this value is ignored (def: Q19)

        Training parameters:
        -a #,#,#        Provide alpha training parameters
        -b #,#,#        Provide beta training parameters
        -d #,#,#        Provide delta training parameters
        -M <file>       Provide a file containing training parameters (will override -a, -b and -d)
                        Values are space-separated:
                        alpha # # #
                        beta # # #
                        delta # # #

        -apply-filter   <yes/no>
        -perl-path      Path to perl installation
        -filter-path    Path to the filter script
        -threshold      THRESHOLD is given, then SNVs will be reported when the selected probability exceeds this
        -c [2|3]                TYPE is the number of classes to consider
                        '2' considers only AA and {AB U BB} (default)
                        '3' considers AA, AB and BB

        -h                              this message

        </code></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/srf2fastq.html" title="class in net.sourceforge.seqware.pipeline.modules">srf2fastq</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.alignment"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/package-summary.html">net.sourceforge.seqware.pipeline.modules.alignment</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/package-summary.html">net.sourceforge.seqware.pipeline.modules.alignment</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/Bfast.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">Bfast</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/BWA.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">BWA</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Align reads via BWA.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/Flat2TableTranscriptDB.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">Flat2TableTranscriptDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given a set reference transcripts in flat files (fasta sequence file & a mapping/assocation file), create & populate
 database tables in UCSC format.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/MakeBigWig.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">MakeBigWig</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Generate bigWig file from a BAM file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/MapSplice.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">MapSplice</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Align reads with via MapSplice

 This module launches a MapSplice alignment of given reads file(s) against given reference
 sequences files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/PrepTranscriptDB.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">PrepTranscriptDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given a set of database tables, construct a reference set of transcripts.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/TrimAdapter.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">TrimAdapter</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trim adapter segment (and any following sequence) from reads.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/TrimAllReads.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">TrimAllReads</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trim all reads down to a user-defined length.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/TrimCountAdapter.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment">TrimCountAdapter</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trim adapter segment (and any following sequence) from reads, and get count & % of reads
 that contain adapter segment.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.alignment.bwa"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/bwa/package-summary.html">net.sourceforge.seqware.pipeline.modules.alignment.bwa</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/bwa/package-summary.html">net.sourceforge.seqware.pipeline.modules.alignment.bwa</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/alignment/bwa/BWA_0_5_9.html" title="class in net.sourceforge.seqware.pipeline.modules.alignment.bwa">BWA_0_5_9</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Align reads via BWA.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.BamFilters"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/package-summary.html">net.sourceforge.seqware.pipeline.modules.BamFilters</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/package-summary.html">net.sourceforge.seqware.pipeline.modules.BamFilters</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/AddRGTags.html" title="class in net.sourceforge.seqware.pipeline.modules.BamFilters">AddRGTags</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Annotate BAM file with read group tags.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/ConvertBAMTranscript2Genome.html" title="class in net.sourceforge.seqware.pipeline.modules.BamFilters">ConvertBAMTranscript2Genome</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Translate reads mapped to transcripts to genomic coordinates.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/RemoveIndelsAndTrim.html" title="class in net.sourceforge.seqware.pipeline.modules.BamFilters">RemoveIndelsAndTrim</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 Module to wrap remove_indels_and_trim script.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/BamFilters/RemoveSingleIndelNotOnEnds.html" title="class in net.sourceforge.seqware.pipeline.modules.BamFilters">RemoveSingleIndelNotOnEnds</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 Module to wrap remove_indels_and_trim script.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.examples"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/examples/package-summary.html">net.sourceforge.seqware.pipeline.modules.examples</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/examples/package-summary.html">net.sourceforge.seqware.pipeline.modules.examples</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/examples/echo.html" title="class in net.sourceforge.seqware.pipeline.modules.examples">echo</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/examples/HelloWorld.html" title="class in net.sourceforge.seqware.pipeline.modules.examples">HelloWorld</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is a hello world module that should give you an idea of how
 to create a module of your own.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.io"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/package-summary.html">net.sourceforge.seqware.pipeline.modules.io</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/package-summary.html">net.sourceforge.seqware.pipeline.modules.io</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/Illumina2SRF.html" title="class in net.sourceforge.seqware.pipeline.modules.io">Illumina2SRF</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This module takes an Illumina Genome Analyzer run folder and converts what
 it finds there into one SRF per lane.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/SRF2Fastq.html" title="class in net.sourceforge.seqware.pipeline.modules.io">SRF2Fastq</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/SRF2Fastq_0_7_2.html" title="class in net.sourceforge.seqware.pipeline.modules.io">SRF2Fastq_0_7_2</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.qc"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/package-summary.html">net.sourceforge.seqware.pipeline.modules.qc</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/package-summary.html">net.sourceforge.seqware.pipeline.modules.qc</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/adapterCont.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">adapterCont</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Basic QC report
 
 This module parses quality report from Illumina pipeline (Gerald)
 
 
  PE Adapters1

  5' P-GATCGGAAGAGCGGTTCAGCAGGAATGCCGAG

  5' ACACTCTTTCCCTACACGACGCTCTTCCGATCT

  PE PCR Primer 1.01

  5' AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT

  3

  PE PCR Primer 2.01

  5' CAAGCAGAAGACGGCATACGAGATCGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT

  PE Read 1 Sequencing Primer

  5' ACACTCTTTCCCTACACGACGCTCTTCCGATCT

  PE Read 2 Sequencing Primer

  5' CGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT
  (1) Normally, PE read 2 is used for SE sequencing and 1st of PE in PE run
  and (2)PE read 1 will is used for 2nd of PE in PE run

  
 Underlying script:  sw_module_adapterCont.pl (by Sara Grimm)
 Newly updated script:  calculate_adapter_contamination.pl    (by Sara Grimm) 6/2/2010               
 Necessary programs:  perl 
  
 Expected output:  adapter_cont.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/alignStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">alignStat</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performing post alignment per lane based counting
  
  
 Underlying script:  sw_module_alignStat.pl
 
 Necessary programs:  perl, java, Picard (SamFormatConverter.jar)
 
 Output:  ~.mapped.txt tab-delimited files in four columns: bam file name, raw read count, 
              mapped read count and aligned percentage (%)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/basicQC.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">basicQC</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module generates pre and post alignment QC reports using
 dqc from the DNAA package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/BCtrend.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">BCtrend</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performing per lane based statistics
  
 Underlying script:  sw_module_BCtrend.pl
                
 Necessary programs:  perl 
  
 Expected output:  read_depth.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/collectFlags.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">collectFlags</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Collecting flag information per lane
  
 Underlying script:  sw_module_extractFlag.pl
                
 Necessary programs:  perl 
  
 Expected output:  flowcell.lane.flagCollection.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/coverageXTranscript.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">coverageXTranscript</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To get coverage across transcript plot
 
 
  
 Underlying script:  sw_module_CoverageAcrossTranscript.pl (by Sara Grimm)
               
 Necessary programs:  perl, path to samtools
  
 Input: alignment file (bam)
        a map file, which is generated by the PrepTranscriptDB module, currently, 
        ~/svnroot/seqware-complete/trunk/seqware-pipeline/data/annotation_reference/hg19_transcripts.hg19.20091027.trmap
 Expected output: coverage across transcript plot</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/dqc.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">dqc</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module generates pre and post alignment QC reports using
 dqc from the DNAA package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/geneCoverage.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">geneCoverage</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To get coverage across transcript plot
 
 
  
 Underlying script:  sw_module_CoverageAcrossTranscript.pl (by Sara Grimm)
               
 Necessary programs:  perl, path to samtools
  
 Input: alignment file (bam)
        a map file, which is generated by the PrepTranscriptDB module, currently, 
        ~/svnroot/seqware-complete/trunk/seqware-pipeline/data/annotation_reference/hg19_transcripts.hg19.20091027.trmap
 Expected output: coverage across transcript plot</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/GenericQCGenome.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">GenericQCGenome</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Estimate % of reads in a given lane that align to a specified set of sequence databases.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/IllQualStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">IllQualStat</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Basic QC report
 
 This module parses quality report from Illumina pipeline (Gerald)
 
 
 Underlying script:  sw_module_parseIllQC.pl
 Necessary programs:  perl 
  
 Expected output:  Ill_QC_summary.csv</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/plotStats.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">plotStats</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Depends on qualStat module, parse sample_stat.txt and 
 generates boxplot sample_bxp.png for base calling scores 
 and per base/cycle % nucleotide distribution sample_nt_distr.png
  
 Underlying script:  /usr/bin/fastq_nucleotide_distribution_graph.sh
                     /usr/bin/fastq_quality_boxplot_graph.sh
                     
 Dependency:  fastx_toolkits from http://hannonlab.cshl.edu/fastx_toolkit/commandline.html
  
 Expected output:  sample_bxp.png & sample_nt_distr.png
 
  *</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/PrepQCgenomes.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">PrepQCgenomes</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Prepare the "QC Genomes" database from external sources.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/QCGenome.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">QCGenome</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To get alignment to QC genome, currenly, human rRNA and viral genome
 
 
  
 Underlying script:  sw_module_RiboAndViralAln.pl (by Sara Grimm)
               
 Necessary programs:  perl path to BWA
  
 Expected output: "mock" alignment to qc genome, text file with key values</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/qualFilter.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">qualFilter</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performing quality flitering on sequence read
  
 Underlying script:  sw_module_qualFilter.pl (by Sara Grimm as  base_quality_scan.pl)
 Necessary programs:  perl 
  
 Expected output:  qual_filter.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/qualStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">qualStat</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performing per lane based statistics
  
 Underlying script:  fastx_quality_stats executable installed
 Dependency:  fastx_quality_stats from http://hannonlab.cshl.edu/fastx_toolkit/commandline.html
              Currently installed at:  /usr/bin/fastx_quality_stats
 Expected output:  qual_filter.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/uniqRead.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">uniqRead</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performing per lane based statistics
  
 Underlying script:  sw_module_readDepth.pl 
                
 Necessary programs:  perl 
  
 Expected output:  read_depth.txt</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.quantification"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/package-summary.html">net.sourceforge.seqware.pipeline.modules.quantification</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/package-summary.html">net.sourceforge.seqware.pipeline.modules.quantification</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/BAM2Count.html" title="class in net.sourceforge.seqware.pipeline.modules.quantification">BAM2Count</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get gene level count from BAM file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/ExonCountsRPKM.html" title="class in net.sourceforge.seqware.pipeline.modules.quantification">ExonCountsRPKM</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculate base counts, coverage, & RPKM per exon given a BAM file of aligned reads and a list of exons.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/GeneCountsRPKM.html" title="class in net.sourceforge.seqware.pipeline.modules.quantification">GeneCountsRPKM</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculate read counts, coverage, & RPKM per transcript and gene given a BAM file of aligned reads.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/reportCardSingle.html" title="class in net.sourceforge.seqware.pipeline.modules.quantification">reportCardSingle</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Generating report card on RNAseq against Microarray - Agilent 244k array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/SpliceJunctionCounts.html" title="class in net.sourceforge.seqware.pipeline.modules.quantification">SpliceJunctionCounts</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculate counts per junction given a BAM file of aligned reads and a list of splice junctions.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.queryengine"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/package-summary.html">net.sourceforge.seqware.pipeline.modules.queryengine</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/package-summary.html">net.sourceforge.seqware.pipeline.modules.queryengine</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/AnnotateVariantDB.html" title="class in net.sourceforge.seqware.pipeline.modules.queryengine">AnnotateVariantDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module uses the VariantConsequenceImporter and
 TagAnnotationImporter from the SeqWare QueryEngine Backend project
 to annotate with coding consequence and dbSNP status.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/AnnotateWithCodingConsequence.html" title="class in net.sourceforge.seqware.pipeline.modules.queryengine">AnnotateWithCodingConsequence</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module uses perl annotation scripts from SeqWare Pipeline to annotate
 with coding consequence status and then generates annotations tag files for
 loading in the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/AnnotateWithDbSNP.html" title="class in net.sourceforge.seqware.pipeline.modules.queryengine">AnnotateWithDbSNP</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module uses an existing variant database, dumps the variants as a BED
 file, uses perl annotation scripts from SeqWare Pipeline to annotate with
 dbSNP status, and create a load file that can be loaded back into the DB.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/CreateVariantDB.html" title="class in net.sourceforge.seqware.pipeline.modules.queryengine">CreateVariantDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module serves to create a variant database from the SeqWare
 Query Engine project that contains SNVs, small indels, and
 coverage.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/queryengine/DumpVariantDB.html" title="class in net.sourceforge.seqware.pipeline.modules.queryengine">DumpVariantDB</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module takes a BerkeleyDB SeqWare Query Engine database and 
 dumps the variants as a series of BED files, one per contig.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.modules.utilities"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/package-summary.html">net.sourceforge.seqware.pipeline.modules.utilities</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Subclasses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/package-summary.html">net.sourceforge.seqware.pipeline.modules.utilities</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/BulkProvisionFiles.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">BulkProvisionFiles</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module takes one or more inputs (S3 URL, HTTP/HTTPS URL, or local file path)
 and copies the file to the specified output (S3 bucket URL, HTTP/HTTPS URL, or local directory path).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/Compress.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">Compress</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:

 This module extends allows for compressing and decompressing files.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/ProvisionDependenciesBundle.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">ProvisionDependenciesBundle</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module takes a software bundle and unzips it to the temporary directory.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/ProvisionFiles.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">ProvisionFiles</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module takes one or more inputs (S3 URL, HTTP/HTTPS URL, or local file
 path) and copies the file to the specified output (S3 bucket URL, HTTP/HTTPS
 URL, or local directory path).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/S3CreateFileURLs.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">S3CreateFileURLs</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module simply lists files at the S3 URL you provide and creates temporary URLs that can be shared 
 with others.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/S3DeleteFiles.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">S3DeleteFiles</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module simply lists files at the S3 URL you provide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/utilities/S3ListFiles.html" title="class in net.sourceforge.seqware.pipeline.modules.utilities">S3ListFiles</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Purpose:
 
 This module simply lists files at the S3 URL you provide.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="net.sourceforge.seqware.pipeline.runner"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
Uses of <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A> in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/runner/package-summary.html">net.sourceforge.seqware.pipeline.runner</A></FONT></TH>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left" COLSPAN="2">Methods in <A HREF="../../../../../../net/sourceforge/seqware/pipeline/runner/package-summary.html">net.sourceforge.seqware.pipeline.runner</A> with parameters of type <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Runner.</B><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/runner/Runner.html#evaluateReturn(net.sourceforge.seqware.pipeline.module.Module, java.lang.String)">evaluateReturn</A></B>(<A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module">Module</A>&nbsp;app,
               java.lang.String&nbsp;methodName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../net/sourceforge/seqware/pipeline/module/Module.html" title="class in net.sourceforge.seqware.pipeline.module"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?net/sourceforge/seqware/pipeline/module//class-useModule.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Module.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
