<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_34) on Mon Oct 22 09:23:56 EDT 2012 -->
<TITLE>
net.sourceforge.seqware.pipeline.modules.qc
</TITLE>

<META NAME="date" CONTENT="2012-10-22">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="net.sourceforge.seqware.pipeline.modules.qc";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?net/sourceforge/seqware/pipeline/modules/qc/package-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<H2>
Package net.sourceforge.seqware.pipeline.modules.qc
</H2>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/adapterCont.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">adapterCont</A></B></TD>
<TD>Basic QC report
 
 This module parses quality report from Illumina pipeline (Gerald)
 
 
  PE Adapters1

  5' P-GATCGGAAGAGCGGTTCAGCAGGAATGCCGAG

  5' ACACTCTTTCCCTACACGACGCTCTTCCGATCT

  PE PCR Primer 1.01

  5' AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT

  3

  PE PCR Primer 2.01

  5' CAAGCAGAAGACGGCATACGAGATCGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT

  PE Read 1 Sequencing Primer

  5' ACACTCTTTCCCTACACGACGCTCTTCCGATCT

  PE Read 2 Sequencing Primer

  5' CGGTCTCGGCATTCCTGCTGAACCGCTCTTCCGATCT
  (1) Normally, PE read 2 is used for SE sequencing and 1st of PE in PE run
  and (2)PE read 1 will is used for 2nd of PE in PE run

  
 Underlying script:  sw_module_adapterCont.pl (by Sara Grimm)
 Newly updated script:  calculate_adapter_contamination.pl    (by Sara Grimm) 6/2/2010               
 Necessary programs:  perl 
  
 Expected output:  adapter_cont.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/alignStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">alignStat</A></B></TD>
<TD>Performing post alignment per lane based counting
  
  
 Underlying script:  sw_module_alignStat.pl
 
 Necessary programs:  perl, java, Picard (SamFormatConverter.jar)
 
 Output:  ~.mapped.txt tab-delimited files in four columns: bam file name, raw read count, 
              mapped read count and aligned percentage (%)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/basicQC.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">basicQC</A></B></TD>
<TD>Purpose:

 This module generates pre and post alignment QC reports using
 dqc from the DNAA package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/BCtrend.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">BCtrend</A></B></TD>
<TD>Performing per lane based statistics
  
 Underlying script:  sw_module_BCtrend.pl
                
 Necessary programs:  perl 
  
 Expected output:  read_depth.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/collectFlags.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">collectFlags</A></B></TD>
<TD>Collecting flag information per lane
  
 Underlying script:  sw_module_extractFlag.pl
                
 Necessary programs:  perl 
  
 Expected output:  flowcell.lane.flagCollection.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/coverageXTranscript.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">coverageXTranscript</A></B></TD>
<TD>To get coverage across transcript plot
 
 
  
 Underlying script:  sw_module_CoverageAcrossTranscript.pl (by Sara Grimm)
               
 Necessary programs:  perl, path to samtools
  
 Input: alignment file (bam)
        a map file, which is generated by the PrepTranscriptDB module, currently, 
        ~/svnroot/seqware-complete/trunk/seqware-pipeline/data/annotation_reference/hg19_transcripts.hg19.20091027.trmap
 Expected output: coverage across transcript plot</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/dqc.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">dqc</A></B></TD>
<TD>Purpose:

 This module generates pre and post alignment QC reports using
 dqc from the DNAA package.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/geneCoverage.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">geneCoverage</A></B></TD>
<TD>To get coverage across transcript plot
 
 
  
 Underlying script:  sw_module_CoverageAcrossTranscript.pl (by Sara Grimm)
               
 Necessary programs:  perl, path to samtools
  
 Input: alignment file (bam)
        a map file, which is generated by the PrepTranscriptDB module, currently, 
        ~/svnroot/seqware-complete/trunk/seqware-pipeline/data/annotation_reference/hg19_transcripts.hg19.20091027.trmap
 Expected output: coverage across transcript plot</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/GenericQCGenome.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">GenericQCGenome</A></B></TD>
<TD>Estimate % of reads in a given lane that align to a specified set of sequence databases.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/IllQualStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">IllQualStat</A></B></TD>
<TD>Basic QC report
 
 This module parses quality report from Illumina pipeline (Gerald)
 
 
 Underlying script:  sw_module_parseIllQC.pl
 Necessary programs:  perl 
  
 Expected output:  Ill_QC_summary.csv</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/plotStats.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">plotStats</A></B></TD>
<TD>Depends on qualStat module, parse sample_stat.txt and 
 generates boxplot sample_bxp.png for base calling scores 
 and per base/cycle % nucleotide distribution sample_nt_distr.png
  
 Underlying script:  /usr/bin/fastq_nucleotide_distribution_graph.sh
                     /usr/bin/fastq_quality_boxplot_graph.sh
                     
 Dependency:  fastx_toolkits from http://hannonlab.cshl.edu/fastx_toolkit/commandline.html
  
 Expected output:  sample_bxp.png & sample_nt_distr.png
 
  *</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/PrepQCgenomes.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">PrepQCgenomes</A></B></TD>
<TD>Prepare the "QC Genomes" database from external sources.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/QCGenome.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">QCGenome</A></B></TD>
<TD>To get alignment to QC genome, currenly, human rRNA and viral genome
 
 
  
 Underlying script:  sw_module_RiboAndViralAln.pl (by Sara Grimm)
               
 Necessary programs:  perl path to BWA
  
 Expected output: "mock" alignment to qc genome, text file with key values</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/qualFilter.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">qualFilter</A></B></TD>
<TD>Performing quality flitering on sequence read
  
 Underlying script:  sw_module_qualFilter.pl (by Sara Grimm as  base_quality_scan.pl)
 Necessary programs:  perl 
  
 Expected output:  qual_filter.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/qualStat.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">qualStat</A></B></TD>
<TD>Performing per lane based statistics
  
 Underlying script:  fastx_quality_stats executable installed
 Dependency:  fastx_quality_stats from http://hannonlab.cshl.edu/fastx_toolkit/commandline.html
              Currently installed at:  /usr/bin/fastx_quality_stats
 Expected output:  qual_filter.txt</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/qc/uniqRead.html" title="class in net.sourceforge.seqware.pipeline.modules.qc">uniqRead</A></B></TD>
<TD>Performing per lane based statistics
  
 Underlying script:  sw_module_readDepth.pl 
                
 Necessary programs:  perl 
  
 Expected output:  read_depth.txt</TD>
</TR>
</TABLE>
&nbsp;

<P>
<DL>
</DL>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/io/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../net/sourceforge/seqware/pipeline/modules/quantification/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?net/sourceforge/seqware/pipeline/modules/qc/package-summary.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
